Several additional ES6 tests

webbedspace opened the pull request:
This closes #98, #123, #133, #140, #141, #142, #143, #144, #145, #146, #147, #148 and #152.

Added rows:
- Proper tail calls
- Shorthand properties
- Shorthand methods
- `__proto__` in object literals (Annex B)
- `Object.getOwnPropertySymbols`
- `Object.prototype.__proto__` (Annex B)
- `Function.prototype.toMethod`
- `String.raw`
- `String.prototype` HTML methods (Annex B)
- `Symbol.create`
- `Symbol.hasInstance`
- `Symbol.isConcatSpreadable`
- `Symbol.isRegExp`
- `Symbol.iterator`
- `Symbol.toPrimitive`
- `Symbol.toStringTag`
- `Symbol.unscopables`
- `RegExp.prototype.match`
- `RegExp.prototype.replace`
- `RegExp.prototype.search`
- `RegExp.prototype.split`
- `RegExp.prototype.compile` (Annex B)
- `Array.prototype.copyWithin`
- `Array.prototype.keys`
- `Array.prototype.values`
- `Array.prototype.entries`
- `Reflect`
- `Reflect.Realm`
- `Reflect.Loader`

Also fixed a few existing tests and updated several source links.

I should note that I did not test RhinoJS, PhantomJS or IE11, but assumed false for all additions (except true for Annex B additions). I couldn't compile EchoJS but inferred support from reading the source. For historical versions of each browser, I used browsershots.com.

Commit:
Made es6 test results whitespace more readable

Commit:
Added test: proper tail calls

Commit:
Altered class test to use extends and static

Traceur and EJS pass this modified test as usual.

Commit:
Added tests: shorthand properties/methods

Commit:
Destructuring test now has object destructuring

Commit:
Added several other tests

 * `__proto__ in object literals`
 * `Object.getOwnPropertySymbols`
 * `Object.prototype.__proto__`
 * `Function.prototype.toMethod`
 * `String.raw`

Commit:
Corrected some test results

Of note: Webkit Nightly now fails the Destructuring test because unlike
Firefox, it can't handle multiple destructurings in a single var
statement.

Commit:
Well-known symbols tests

In addition to checking for the well-known symbol properties on Symbol,
this replaces the broken Iterators test with the Symbol.iterator test.

webbedspace committed:
Altered some links to point to the ES6 spec

webbedspace committed:
Corrected the proper tail call test

The previous test wasn't pushing the upper limit of each engine's call stack, and was causing false positives. All test results are now set to 'no' pending further examination.

webbedspace committed:
Updated various ES6 reference links

Generally, I prioritised linking to the specification for individual
methods/properties, but not for syntax structures with no specific
descriptive section in the document.

webbedspace committed:
Added test: Array.prototype.copyWithin

webbedspace committed:
Added note to WebKit destructuring support

Decided to mark it as "true" but with the multiple destructuring bug
acknowledged in a note.

webbedspace committed:
ES6 tail call test is too hard

If that actually ran 5e6 iterations, it would block for >1 minute. ;_;

webbedspace committed:
Updated Symbol tests to check for functionality

I read the ES6 spec for each well-known symbol and hope these tests are
sufficient to determine at least partial functionality.

webbedspace committed:
Added note for __proto__ in object literals

webbedspace committed:
ES6: Added RegExp.prototype method tests

webbedspace committed:
ES6: Added Array.prototype iterator methods

* Array.prototype.entries
* Array.prototype.keys
* Array.prototype.values

Historical Chrome data is approximate based on this Chromium commit:
https://code.google.com/p/v8/source/detail?r=15620

webbedspace committed:
ES6: added String HTML methods

These annex B methods, dating back to Netscape Navigator, were
non-standard prior to ES6.

webbedspace committed:
ES6: Added RegExp.prototype.compile test

webbedspace committed:
Fixed bug in ES6 class test

webbedspace committed:
ES6: Added Reflect tests

These are separate tests for Reflect, Reflect.Realm and Reflect.Loader,
granted that the Reflect object's own methods are intended for a
different sort of use than the Realm or Module objects.

kangax commented:
@webbedspace thanks for this massive effort! Pulled locally and it looks good; the only thing — FF33 reports `true` on shortand properties for me, but I can update it already after.

@rwaldron @domenic any feedback before I merge this in?

domenic commented:
Everything looks good to me!

webbedspace committed:
Added more destructuring tests

Added:
* Destructuring parameters
* Destructuring defaults
* Destructuring with rest
* Update to the basic test to check for nested structures, aliasing,
and skipping array elements.

Also, unrelatedly added:
* Symbol.for() and Symbol.keyFor() test
* Array#[Symbol.unscopables] test
* Corrections to a few test results.

Thanks to @rwaldron for feedback.

webbedspace committed:
Fixed PhantomJS result for Object#__proto__

PH's __proto__ implementation does not have a visible property
descriptor.

Also decided to add guards for the ES5 methods used in the test. Since
Object.getOwnPropertyDescriptor is the only way to guarantee that it is
implemented as a get/set property, we might as well fail the test if
it's absent. (Incidentally, everything that fully supports
Object.getOwnPropertyDescriptor also supports Object.create.)

webbedspace committed:
Updated destructuring tests: non-matches

These tests now check that non-matching destructuring
patterns/properties create variables set to undefined (or an empty
array, in the case of rests).

Also updated the link for the global symbol registry test.

webbedspace committed:
Added extra check to "__proto__ in literals" test

To be fully compliant, the __proto__ syntax should not apply when
"__proto__" is a computed property name.

webbedspace committed:
Made Array#[Symbol.unscopables] forward-compatible

If ES7+ adds Array#contains or Array#has or something else, then those
might be added to the unscopables blacklist. (Even then I'm assuming
that the order and consecutiveness of the ES6 unscopables remains fixed
for all eternity.)
